apiserver/common/tools.go

```diff
@@ -230,182 +243,52 @@ type ToolsFinder interface {
 }
 
 type toolsFinder struct {
-	controllerConfigService ControllerConfigService
-	toolsStorageGetter      ToolsStorageGetter
-	urlGetter               ToolsURLGetter
-	newEnviron              NewEnvironFunc
-	store                   objectstore.ObjectStore
+	agentFinderService AgentFinderService
+	toolsStorageGetter ToolsStorageGetter
+	urlGetter          ToolsURLGetter
+	newEnviron         NewEnvironFunc
+	store              objectstore.ObjectStore
 }
 
 // NewToolsFinder returns a new ToolsFinder, returning tools
 // with their URLs pointing at the API server.
 func NewToolsFinder(
-	controllerConfigService ControllerConfigService,
+	agentFinder AgentFinderService,
 	toolsStorageGetter ToolsStorageGetter,
 	urlGetter ToolsURLGetter,
 	newEnviron NewEnvironFunc,
 	store objectstore.ObjectStore,
 ) *toolsFinder {
 	return &toolsFinder{
-		controllerConfigService: controllerConfigService,
-		toolsStorageGetter:      toolsStorageGetter,
-		urlGetter:               urlGetter,
-		newEnviron:              newEnviron,
-		store:                   store,
+		agentFinderService: agentFinder,
+		toolsStorageGetter: toolsStorageGetter,
+		urlGetter:          urlGetter,
+		newEnviron:         newEnviron,
+		store:              store,
 	}
 }
 
 // FindAgents calls findMatchingTools and then rewrites the URLs
 // using the provided ToolsURLGetter.
 func (f *toolsFinder) FindAgents(ctx context.Context, args FindAgentsParams) (coretools.List, error) {
-	list, err := f.findMatchingAgents(ctx, args)
```

Comment:
This is the essence of the change - the generic logic in this FindAgents method which uses the provider is moved to a FindAgents method on the modelagent service where the provider can be accessed via the tracker.